core/state.py

```diff
@@ -2309,6 +2309,59 @@ def PopContextStack(self):
         return self.ctx_stack.pop()
 
 
+class Procs:
+
+    def __init__(self, mem):
+        # type: (Mem) -> None
+        self.mem = mem
+        self.procs = {}  # type: Dict[str, value.Proc]
+
+    def SetProc(self, name, proc):
+        # type: (str, value.Proc) -> None
+        self.mem.var_stack[0][name] = Cell(False, False, False, proc)
+
+    def SetShFunc(self, name, proc):
+        # type: (str, value.Proc) -> None
+        self.procs[name] = proc
+
+    def Get(self, name):
+        # type: (str) -> value.Proc
+        vars = self.mem.var_stack[0]
+        if name in vars:
+            maybe_proc = vars[name]
+            if maybe_proc.val.tag() == value_e.Proc:
+                return cast(value.Proc, maybe_proc.val)
+
+        if name in self.procs:
+            return self.procs[name]
+
+        return None
+
+    def Del(self, to_del):
```

Comment:
Is this function only used by `unset -f` ?  If so then maybe it can only call `mylib.dict_erase()` on `self.sh_funcs` ?

I think the way to delete a proc should now be the same way as we delete any other variable.

----

Or I guess if it makes tests fail, we could do this in a separate step

Reply:
You're right, this can be simplified to just `mylib.dict_erase()`.